### PR TITLE
New `matekf411rx_fs` target which implements FlySky AFHDS and AFHDS2A…

### PR DESCRIPTION
**Pull request by sakitume**
… protocol support

Developed and tested on CrazybeeF4 Lite FS and CrazybeeF3 FS flight controllers

### CODE REVIEW

File: src/rx/rx_flysky_hal.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  }
+}
+
+//------------------------------------------------------------------------------
+void flysky_fast_blink_rx_led() {
+  blink_rx_led(250);
+}
+
+//------------------------------------------------------------------------------
+// TODO: @hanfer is there something equivalent already in Quicksilver?
```
Comment (bkleiner) [inline src/rx/rx_flysky_hal.c:76]: No, but seems useful. probably shouldn't live here tho. is there a more cross-mcu way of reading these? 
Edit: jup there is, seems a define `UID_BASE` with the address is provided in the mcu headers. 
lets move this function to src/util/ somewhere

File: src/rx/rx_flysky_hal.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+//------------------------------------------------------------------------------
+static void blink_rx_led(uint32_t blink_interval_ms) {
```
Comment (bkleiner) [inline src/rx/rx_flysky_hal.c:47]: _usually_ we just define the RX led as another system led. so maybe remove this and use the existing led functions with `LED_ALL` not like you are doing much else while you are binding.

File: src/rx/rx_flysky_hal.c [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  uint32_t now = time_millis();
+  if (now >= (last_time + blink_interval_ms)) {
+    last_time = now;
+    gpio_pin_toggle(RX_LED_PIN);
+  }
+}
+
+//------------------------------------------------------------------------------
+// If failsafe the then slow blink, otherwise keep led steady (off)
+void flysky_update_rx_led(bool failsafe) {
```
Comment (bkleiner) [inline src/rx/rx_flysky_hal.c:58]: regular led logic will handle failsafe too, lets remove this one.

File: src/rx/rx_flysky_hal.c [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+//------------------------------------------------------------------------------
+// This file contains hardware related functions and is kept separate
+// from the rest of the RX code so that the RX code can be developed/used/tested
+// with other projects/frameworks
+
+//------------------------------------------------------------------------------
+// If 'init' is true, this records state info about the bind button pin so we can
+// later debounce and detect state changes. Otherwise it checks the state
+// (performing a debounce check) and returns true if button was depressed
+bool flysky_check_bind_button(bool init) {
```
Comment (bkleiner) [inline src/rx/rx_flysky_hal.c:17]: historically we do not use bind buttons. personally i hate them, feels like something that can just be handled so much more convenient via other means, but if are to introduce them, this should be a system for all protocols.
please eliminate and if we want to change the behaviour here, let's revisit in another pr/whatever.

File: src/targets/matekf411rx_fs/target.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+// PB9 is used as output (push/pull, no pullup/pulldown) for RX LED
+// PB2 is used as input (no pullup/pulldown) for RX bind pushbutton
+#define USE_A7105
+#define A7105_SPI_PORT SPI_PORT3
+#define A7105_NSS_PIN PIN_A15
+#define A7105_GIO1_PIN PIN_A14    // EXTI15_10_IRQHandler
+#define RX_LED_PIN PIN_B9
+#define RX_BIND_PIN PIN_B2
+
+#ifdef SERIAL_RX
```
Comment (bkleiner) [inline src/targets/matekf411rx_fs/target.h:45]: deprecated, lets drop this including the RX_USART assigment

File: src/flash.h [hunk 6]
```diff
@@ -29,6 +30,16 @@ typedef struct {
     rx_bayang_bind_data_t bayang;
     rx_unified_bind_data_t unified;
     rx_elrs_bind_data_t elrs;
+
+// TODO: For some reason adding this to the union actually changes the union size!
```
Comment (bkleiner) [inline src/flash.h:34]: that is.....perplexing.....brb, digging in the c standard

**Review by bkleiner — CHANGES_REQUESTED** [review]
Overall this looks great, well done, very neat.

as for the code style, with the platformio extension for vscode, should have come a c/c++ extension which will pick up the clang-format file in the root of the repository

by adding to your vscode settings, clang format will run on every save, ensuring the codestyle is consistent across the codebase.
```
"editor.formatOnSave": true,
```

please also add/update the respective targets in the script/target.json file, so they are picked up by the ci.

File: src/flash.h (hunk 6 above)
Comment (bkleiner) [reply src/flash.h:34]: https://github.com/BossHobby/QUICKSILVER/commit/5b83a1b86dcc613425232b6a2bedaab53824a9f0 should fix this issue

File: platformio.ini [hunk 7]
```diff
@@ -251,6 +251,12 @@ build_flags =
   ${stm32f411.build_flags}
   -Isrc/targets/matekf411rx
 
+[env:matekf411rx_fs]
```
[the remark this answers is not in the document]
Comment (sakitume) [reply platformio.ini]: Addressed with 993df7e

File: src/drivers/drv_spi_a7105.c [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+static void a7105_write_multi(uint8_t reg, const uint8_t *data, uint8_t len) {
+  spi_txn_t *txn = spi_txn_init(&bus, NULL);
+  spi_txn_add_seg_const(txn, reg);
+  spi_txn_add_seg(txn, NULL, data, len);
+  spi_txn_submit(txn);
+
+  // needs to sync ???
+  spi_txn_continue_ex(&bus, true);
```
[the remark this answers is not in the document]
Comment (sakitume) [reply src/drivers/drv_spi_a7105.c:45]: Using `true` or `false` with some very limited testing doesn't seem to affect the behavior. Since I've been using it like this for a while I'd like to retain this code as it is.

*But* I would like to review the entire SPI implementation a bit more as I used the FrSky (CC2500) code as the basis for this source file rather than taking a deep dive into understanding how the SPI implementation works. It seems it can be blocking or not depending on how you call into it.

In my test project (outside of Quicksilver) I use a soft spi approach (bit bang) and it works great. My link quality for AFHDS is rock solid (666 packets per second). But in Quicksilver it fluctuates every now and then and the only real differences between my test project and Quicksilver is possibly the looptime (I know my test project is rock solid 250us) and the soft spi (which is synchronous).

tl;dr I want to revisit this SPI implementation with your help but hope we can accept this for now

File: src/drivers/drv_spi_a7105.c [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+static void a7105_read_multi(uint8_t reg, uint8_t *result, uint8_t len) {
+  spi_txn_t *txn = spi_txn_init(&bus, NULL);
+  spi_txn_add_seg(txn, &reg, &reg, 1);
+  spi_txn_add_seg(txn, result, NULL, len);
+  spi_txn_submit_wait(&bus, txn);
+}
+
+void a7105_strobe(a7105_strobe_t address)
+{
```
[the remark this answers is not in the document]
Comment (sakitume) [reply src/drivers/drv_spi_a7105.c]: This should be addressed with 993df7e

File: src/rx/rx_flysky_afhds.c [hunk 10]
```diff
@@ -0,0 +1,111 @@
+#include "rx_flysky.h"
+
+#if defined(RX_FLYSKY)
+#include "drv_spi_a7105.h"
+#include <string.h>
+
+//------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (sakitume) [reply src/rx/rx_flysky_afhds.c:7]: I can see how this may seem distracting, but when I develop various routines I like to delimit functions and provide function header comments of some sort when it makes sense. I'd like to retain these for now in case I need to get back to this code and perform maintenance. 

In the future I suppose we can remove them to maintain consistency with the rest of the codebase.

File: src/rx/rx_flysky_hal.c (hunk 1 above)
Comment (sakitume) [reply src/rx/rx_flysky_hal.c]: See c7d965a

Remove `get_chip_id()` from `rx_flysky_hal.c` and replace it with new `get_chip_uid()` found in `util/util.c`

File: src/rx/rx_flysky_hal.c (hunk 2 above)
Comment (sakitume) [reply src/rx/rx_flysky_hal.c]: Now defining LED2PIN for the RX led so it behaves as a system LED. See e81d403

File: src/rx/rx_flysky_hal.c (hunk 3 above)
Comment (sakitume) [reply src/rx/rx_flysky_hal.c]: Now that RX led is behaving as a system led, and using the led behavior system to support blinking bind/failsafe, we're able to remove all of the led management routines. See e81d403

File: src/rx/rx_flysky_hal.c (hunk 4 above)
Comment (sakitume) [reply src/rx/rx_flysky_hal.c]: Happy to remove bind button support. Removed with e81d403

File: src/targets/matekf411rx_fs/target.h (hunk 5 above)
Comment (sakitume) [reply src/targets/matekf411rx_fs/target.h]: Removed RX_USART definition with 993df7e

File: src/flash.h (hunk 6 above)
Comment (sakitume) [reply src/flash.h]: Thank you. I updated my `develop` branch to obtain your fix, then rebased this feature branch on top of that. Then I
removed workaround code for bind storage data type weirdness. The `rx_flysky_bind_data_t` type member of the anonymous union in `rx_bind_storage_t` is now enabled.

See eadd532

File: src/drivers/drv_spi_a7105.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (bkleiner) [reply src/drivers/drv_spi_a7105.c:45]: yeah, i would bet the fluctuations are down to a loop taking longer and it not returning to rx code in time. We got a lot of other stuff going on at the same time. i have been working on a scheduler https://github.com/BossHobby/QUICKSILVER/tree/feature-scheduler to ensure crictical paths, like the rx code, are always ran on time.
